File: jambeth/jAmbeth.Merge/src/main/java/de/osthus/ambeth/merge/ValueObjectConfig.java
package de.osthus.ambeth.merge;

import java.util.HashSet;
import java.util.Set;

import de.osthus.ambeth.collections.HashMap;
import de.osthus.ambeth.collections.IMap;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;

public class ValueObjectConfig implements IValueObjectConfig
{
	@SuppressWarnings("unused")
	@LogInstance(ValueObjectConfig.class)
	private ILogger log;

	protected Class<?> entityType;

	protected Class<?> valueType;

	protected Set<String> listTypeMembers = new HashSet<String>();

	protected IMap<String, ValueObjectMemberType> memberTypes = new HashMap<String, ValueObjectMemberType>();

	protected IMap<String, Class<?>> collectionMemberTypes = new HashMap<String, Class<?>>();

	protected IMap<String, String> boToVoMemberNameMap = new HashMap<String, String>();

	@Override
	public Class<?> getEntityType()
	{
		return entityType;
	}

	public void setEntityType(Class<?> entityType)
	{
		this.entityType = entityType;
	}

	@Override
	public Class<?> getValueType()
	{
		return valueType;
	}

	public void setValueType(Class<?> valueType)
	{
		this.valueType = valueType;
	}

	@Override
	public String getValueObjectMemberName(String businessObjectMemberName)
	{
		String voMemberName = boToVoMemberNameMap.get(businessObjectMemberName);
		if (voMemberName == null)
		{
			return businessObjectMemberName;
		}
		else
		{
			return voMemberName;
		}
	}

	public void putValueObjectMemberName(String businessObjectMemberName, String valueObjectMemberName)
	{
		if (!boToVoMemberNameMap.putIfNotExists(businessObjectMemberName, valueObjectMemberName))
		{
			throw new IllegalStateException("Mapping for member '" + businessObjectMemberName + "' already defined");
		}
	}

	@Override
	public boolean holdsListType(String memberName)
	{
		return listTypeMembers.contains(memberName);
	}

	public void addListTypeMember(String memberName)
	{
		listTypeMembers.add(memberName);
	}

	@Override
	public ValueObjectMemberType getValueObjectMemberType(String valueObjectMemberName)
	{
		ValueObjectMemberType memberType = memberTypes.get(valueObjectMemberName);
		if (memberType == null)
		{
			memberType = ValueObjectMemberType.UNDEFINED;
		}
		return memberType;
	}

	public void setValueObjectMemberType(String valueObjectMemberName, ValueObjectMemberType memberType)
	{
		if (!memberTypes.putIfNotExists(valueObjectMemberName, memberType))
		{
			throw new IllegalStateException("Type entry for member '" + valueObjectMemberName + "' already exists");
		}
	}

	@Override
	public boolean isIgnoredMember(String valueObjectMemberName)
	{
		ValueObjectMemberType memberType = getValueObjectMemberType(valueObjectMemberName);
		return memberType == ValueObjectMemberType.IGNORE;
	}

	@Override
	public Class<?> getMemberType(String memberName)
	{
		return collectionMemberTypes.get(memberName);
	}

	public void putMemberType(String memberName, Class<?> elementType)
	{
		if (!collectionMemberTypes.putIfNotExists(memberName, elementType))
		{
			throw new IllegalStateException("Type for member '" + memberName + "' already defined");
		}
	}
}

File: ambeth/Ambeth.Merge/ambeth/merge/ValueObjectConfig.cs
using System;
using System.Collections.Generic;
using De.Osthus.Ambeth.Util;

namespace De.Osthus.Ambeth.Merge
{
    public class ValueObjectConfig : IValueObjectConfig
    {
        public Type EntityType { get; set; }

        public Type ValueType { get; set; }

        protected ISet<String> ignoredMembers = new HashSet<String>();

        protected ISet<String> listTypeMembers = new HashSet<String>();

        protected IDictionary<String, ValueObjectMemberType?> valueObjectMemberTypes = new Dictionary<String, ValueObjectMemberType?>();

        protected IDictionary<String, Type> memberTypes = new Dictionary<String, Type>();

        protected IDictionary<String, String> boToVoMemberNameMap = new Dictionary<String, String>();

        public String GetValueObjectMemberName(String businessObjectMemberName)
        {
            String voMemberName = DictionaryExtension.ValueOrDefault(boToVoMemberNameMap, businessObjectMemberName);
            if (voMemberName == null)
            {
                return businessObjectMemberName;
            }
            else
            {
                return voMemberName;
            }
        }

        public void PutValueObjectMemberName(String businessObjectMemberName, String valueObjectMemberName)
        {
            if (!boToVoMemberNameMap.ContainsKey(businessObjectMemberName))
            {
                boToVoMemberNameMap.Add(businessObjectMemberName, valueObjectMemberName);
            }
            else
            {
                throw new InvalidOperationException("Mapping for member '" + businessObjectMemberName + "' already defined");
            }
        }

        public bool HoldsListType(String memberName)
        {
            return listTypeMembers.Contains(memberName);
        }

        public void AddListTypeMember(String memberName)
        {
            listTypeMembers.Add(memberName);
        }

        public ValueObjectMemberType GetValueObjectMemberType(String valueObjectMemberName)
        {
            ValueObjectMemberType? memberType = DictionaryExtension.ValueOrDefault(valueObjectMemberTypes, valueObjectMemberName);
            if (memberType == null)
            {
                memberType = ValueObjectMemberType.UNDEFINED;
            }
            return memberType.Value;
        }

        public void SetValueObjectMemberType(String valueObjectMemberName, ValueObjectMemberType memberType)
        {
            if (valueObjectMemberTypes.ContainsKey(valueObjectMemberName))
            {
                throw new Exception("Type entry for member '" + valueObjectMemberName + "' already exists");
            }
            valueObjectMemberTypes.Add(valueObjectMemberName, memberType);
        }

        public bool IsIgnoredMember(String valueObjectMemberName)
        {
            ValueObjectMemberType memberType = GetValueObjectMemberType(valueObjectMemberName);
            return memberType == ValueObjectMemberType.IGNORE;
        }

        public Type GetMemberType(String memberName)
        {
            return DictionaryExtension.ValueOrDefault(memberTypes, memberName);
        }

        public void PutMemberType(String memberName, Type elementType)
        {
            if (!memberTypes.ContainsKey(memberName))
            {
                memberTypes.Add(memberName, elementType);
            }
            else
            {
                throw new InvalidOperationException("Type for member '" + memberName + "' already defined");
            }
        }
    }
}
